### PR TITLE
Core refactoring

### PR DESCRIPTION
**Pull request by mephenor**
###  Changes

- Split core interrogator functionality into more dedicated functions
- Split data passing through interrogation functionality into `InterrogationSubject` and `ProcessingResult` container models to enable slimmer function signatures
- Removed core dependency on `ghga_event_schemas` by use of the `InterrogationSubject` model
- Outgoing events are now consistently fired after all internal processing except storing the fingerprint is done
- Distinguish between exceptions concerning the file being processed and exceptions from misconfiguration, network or EKSS issues that can be recovered from - in the sense that no validation failure should be announced. These should take down this service instead to trigger root cause investigation
- Fingerprinting is now also done for failed processing if the failure occurs during checksum comparison or re-encryption, i.e. only transient errors in the re-encryption and staging functionality should not store a fingerprint.
- Added more logging

### CODE REVIEW

File: src/irs/adapters/outbound/http/exceptions.py
Comment (TheByronHimes) [inline src/irs/adapters/outbound/http/exceptions.py:1]: KnownError should either be disposed of or renamed to something more indicative of EnvelopeDecryptionError and MalformedOrMissingEnvelope.

File: src/irs/core/models.py [hunk 1]
```diff
@@ -41,9 +40,22 @@ class StagingObject(BaseModel):
...
+    decrypted_size: int
+    expected_decrypted_sha256: str
+    upload_date: str
+    submitter_public_key: str
+
+
 class UploadReceivedFingerprint(BaseModel):
     """
-    Stores the hash of a FileUploadReceived payload along with its generation
+    Stores the hash of a InterrogationSubject payload along with its generation
```
Comment (TheByronHimes) [inline src/irs/core/models.py]:
```suggestion
    Stores the hash of an InterrogationSubject payload along with its generation
```

File: src/irs/core/interrogator.py [hunk 2]
```diff
@@ -214,3 +247,4 @@ async def remove_staging_object(self, *, file_id: str, storage_alias: str) -> No
             bucket_id=staging_bucket_id, object_id=staging_object.object_id
         )
         await self._staging_object_dao.delete(id_=file_id)
+        log.debug("Successfullly removed staging object for file '%s'.", file_id)
```
Comment (TheByronHimes) [inline src/irs/core/interrogator.py]:
```suggestion
        log.debug("Successfully removed staging object for file '%s'.", file_id)
```

File: src/irs/core/storage_inspector.py [hunk 3]
```diff
@@ -55,6 +55,62 @@ def __init__(
         self._staging_object_dao = staging_object_dao
         self._object_storages = object_storages
 
+    async def _inspect_object(
+        self, *, bucket_id: str, object_id: str, storage_alias: str
+    ):
+        """Check one specific object in a specified storage node."""
+        try:
```
Comment (TheByronHimes) [inline src/irs/core/storage_inspector.py]:
```suggestion
        """Check one specific object in a specified storage node."""
        extra = {
            "object_id": object_id,
            "bucket_id": bucket_id,
            "storage_alias": storage_alias,
        }
        
        try:
```
Then you could add the file_id in the case of a stale object

File: src/irs/core/storage_inspector.py [hunk 4]
```diff
@@ -55,6 +55,62 @@ def __init__(
...
+            )
+        except NoHitsFoundError:
+            # This can happen in one of two cases:
+            # 1) Failed to send success event
+            # 2) Success event sent, but failed to insert staging object entry
+            extra = {
+                "object_id": object_id,
+                "bucket_id": bucket_id,
+                "storage_alias": storage_alias,
+            }
```
Comment (TheByronHimes) [inline src/irs/core/storage_inspector.py]:
```suggestion
```

File: src/irs/core/storage_inspector.py [hunk 5]
```diff
@@ -55,6 +55,62 @@ def __init__(
...
+            )
+            return
+
+        except MultipleHitsFoundError as error:
+            # can only happen if the same object ID is generated from uuid4()
+            extra = {
+                "object_id": object_id,
+                "bucket_id": bucket_id,
+                "storage_alias": storage_alias,
+            }
```
Comment (TheByronHimes) [inline src/irs/core/storage_inspector.py]:
```suggestion
```

File: src/irs/core/storage_inspector.py [hunk 6]
```diff
@@ -55,6 +55,62 @@ def __init__(
...
+            minutes=self._config.object_stale_after_minutes
+        )
+        if staging_object.creation_date <= stale_as_of:
+            # only log for now, but this points to an underlying issue
+            extra = {
+                "object_id": object_id,
+                "file_id": staging_object.file_id,
+                "bucket_id": bucket_id,
+                "storage_alias": storage_alias,
+            }
```
Comment (TheByronHimes) [inline src/irs/core/storage_inspector.py]:
```suggestion
            extra["file_id"] = staging_object.file_id
```

**Review by TheByronHimes — CHANGES_REQUESTED** [review]
On the whole looks like a good refactor, just some small items to address

File: src/irs/adapters/outbound/http/exceptions.py
Comment (mephenor) [reply src/irs/adapters/outbound/http/exceptions.py:1]: It's an `EnvelopeError` now

**Review by TheByronHimes — APPROVED** [review]

**Pull request merged**